build_tools/github_actions/should_ci_run.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""Determines whether CI should run on a given PR.
+
+Exit code 0 indicates that it should and exit code 2 indicates that it should
+not.
+"""
+
+import fnmatch
+import os
+import subprocess
+import sys
+
+SKIP_CI_LABEL = "skip-ci"
+
+# Note that these are fnmatch patterns, which are not the same as gitignore
+# patterns because they don't treat '/' specially. The standard library doesn't
+# contain a function for gitignore style "wildmatch". There's a third-party
+# library pathspec (https://pypi.org/project/pathspec/), but it doesn't seem
+# worth the dependency.
+SKIP_PATH_PATTERNS = [
+    "docs/*",
+    "experimental/*",
+    "build_tools/kokoro/*",
+    "build_tools/buildkite/*",
+    ".github/ISSUE_TEMPLATE/*",
+    "*.cff",
+    "*.clang-format",
+    "*.git-ignore",
+    "*.md",
+    "*.natvis",
+    "*.pylintrc",
+    "*.rst",
+    "*.toml",
+    "*.yamllint.yml",
+    "*.yapf",
+    "*CODEOWNERS",
+    "*AUTHORS",
+    "*LICENSE",
+]
+
+
+def skip_path(path):
+  return any(fnmatch.fnmatch(path, pattern) for pattern in SKIP_PATH_PATTERNS)
+
+
+def get_modified_paths(base_ref):
+  return subprocess.run(["git", "diff", "--name-only", base_ref],
+                        stdout=subprocess.PIPE,
+                        check=True,
+                        text=True,
+                        timeout=20).stdout.splitlines()
```

Comment:
What happens if this timeout is exceeded? Can you catch that and default to running CI?

Reply:
It will raise a `TimeoutExpired` which will propagate up to the top level and cause the script to exit with code 1. Then the job running this will fail. I put the timeout in here as just a general good practice to not let things run forever. The thought was that if this command takes longer than this then something has gone wrong and we should just rerun the job. I guess it's possible for the diff to be so large that this takes a really long time. In that case just running the CI anyway seems reasonable. I think we might want to give it longer though. Thoughts?

Reply:
Yeah, I figure that large diffs or any filesystem lag / flakes should just run the CI. I'd expect that failures here would be transient and would not indicate major errors in the rest of the pipeline.

Reply:
Yeah I'd expect that they'd be transient, but running this is way faster than running the rest of the CI pipeline, so it's an expensive side to err on. OTOH, most PRs should probably be running the CI. I'll extend the timeout and make it silence this error.